wallet/mnemonic.js

```diff
@@ -8,13 +12,85 @@ function generateMnemonic() {
 
 function validateMnemonic(mnemonic) {
   try {
-    return !!mnemonic && bip39.validateMnemonic(mnemonic)
+    return !!mnemonic && (bip39.validateMnemonic(mnemonic) || isPaperWalletMnemonic(mnemonic))
   } catch (e) {
     return false
   }
 }
 
+function isPaperWalletMnemonic(mnemonic) {
+  return mnemonicToList(mnemonic).length === 27
+}
+
+function mnemonicToList(mnemonic) {
+  return mnemonic.split(' ')
+}
+
+function mnemonicToHashSeed(mnemonic) {
+  if (!validateMnemonic(mnemonic)) {
+    const e = new Error('Invalid or unsupported mnemonic format')
+    e.name = 'InvalidArgumentException'
+    throw e
+  }
+
+  if (isPaperWalletMnemonic(mnemonic)) {
+    mnemonic = decodePaperWalletMnemonic(mnemonic)
+  }
+
+  const ent = Buffer.from(bip39.mnemonicToEntropy(mnemonic), 'hex')
+
+  return cbor.encode(hashBlake2b256(cbor.encode(ent)))
+}
+
+function mnemonicToPaperWalletPasphrase(mnemonic, password) {
+  const mnemonicBuffer = Buffer.from(mnemonic, 'utf8')
+  const salt = `mnemonic${password || ''}`
+  const saltBuffer = Buffer.from(salt, 'utf8')
+  return pbkdf2Sync(mnemonicBuffer, saltBuffer, 2048, 32, 'sha512').toString('hex')
+}
+
+function decodePaperWalletMnemonic(paperWalletMnemonic) {
+  const paperWalletMnemonicAsList = mnemonicToList(paperWalletMnemonic)
+
+  if (paperWalletMnemonicAsList.length !== 27) {
+    throw Error(
+      `Paper Wallet Mnemonic must be 27 words, got ${paperWalletMnemonicAsList.length} instead`
+    )
+  }
+
+  const scrambledPart = paperWalletMnemonicAsList.slice(0, 18).join(' ')
+  const passphrasePart = paperWalletMnemonicAsList.slice(18, 27).join(' ')
+
+  const passphrase = mnemonicToPaperWalletPasphrase(passphrasePart)
```

Comment:
why is there injected `passphrasePart` in the spot where the mnemonic should be? I would rename `passphrasePart` to mnemonic, or if it is not a mnemonic then I would suggest to rename function `mnemonicToPaperWalletPasphrase` so it will be clear what it is for.